Add OTel for RUM docs (#4115)
elastic/docs-content: solutions/observability/applications/otel-rum.md

While this guide uses upstream OpenTelemetry instrumentation, you can also use the [EDOT Collector](elastic-agent://reference/edot-collector/index.md) components as part of your data ingestion pipeline.

:::{warning}
Avoid using OpenTelemetry alongside any other {{apm-agent}}, including Elastic {{product.apm}} agents. Running multiple agents in the same application process might lead to conflicting instrumentation, duplicate telemetry, or other unexpected behavior.
:::

## Prerequisites

This guide assumes you're using an {{product.observability}} deployment. You can use an existing one or set up a new one. If you're new to {{product.observability}}, follow the guidelines in [Get started with {{product.observability}}](/solutions/observability/get-started.md).

### OTLP endpoint

> You have two main options for setting up an OTLP endpoint:

Reviewer: At the end of the page we are getting back to this topic. Should we merge the two into one?

#### Use an existing OpenTelemetry Collector [use-an-existing-open-telemetry-collector]

If you already have an OpenTelemetry Collector in your infrastructure, you can reuse it to ingest the traces, metrics, and logs generated by your instrumented web applications.

::::{tab-set}

:::{tab-item} Reverse proxy (recommended)

If your Collector is not publicly available or you want to control where the data is coming from, use a reverse proxy to forward data from the browsers through your web application. This is the recommended method.

The following snippet shows the configuration for an NGINX reverse proxy to forward all telemetry to the Collector located at `collector.example.com` from the origin `webapp.example.com`:

```nginx
server {
    listen 80 default_server;
    server_name _;
    location / {
        # Take care of preflight requests
        if ($request_method = 'OPTIONS') {
            add_header 'Access-Control-Max-Age' 1728000;
            add_header 'Access-Control-Allow-Origin' 'webapp.example.com' always;

>             add_header 'Access-Control-Allow-Headers' 'Accept,Accept-Language,Content-Language,Content-Type' always;

Reviewer: Do we need to explicitly allow the `traceparent,tracestate,baggage` headers? Same question for line 69.

Author: +CC @david-luna

Reviewer 2: nope. The reason is because export requests are not instrumented. These headers should be added only in the requests to downstream services.

            add_header 'Access-Control-Allow-Credentials' 'true' always;
            add_header 'Content-Type' 'text/plain charset=UTF-8';
            add_header 'Content-Length' 0;
            return 204;
        }

        add_header 'Access-Control-Allow-Origin' 'webapp.example.com' always;
        add_header 'Access-Control-Allow-Credentials' 'true' always;
        add_header 'Access-Control-Allow-Headers' 'Accept,Accept-Language,Content-Language,Content-Type' always;
        proxy_http_version 1.1;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        
        proxy_pass https://collector.example.com:4318;
    }
}
```
:::

:::{tab-item} Public Collector with CORS

> If the Collector is publicly available, you can send the telemetry data directly to it. Your Collector should be available under a domain name, for example `collector.example.com:443`. Your web application should send data from its own origin `webapp.example.com` to a different one, and [Cross-Origin Resource Sharing](https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/CORS) (CORS) should be configured so browsers allow sending data to a different origin.

Reviewer: The collector endpoint config looks like this for TLS:

```
    receivers:
      # Receives data from other Collectors in Agent mode
      otlp:
        protocols:
          grpc:
            endpoint: 0.0.0.0:4317 # Listen on all interfaces
           # tls:
             # cert_file: "/etc/ssl/certs/collector-server.crt"
             # key_file: "/etc/ssl/private/collector-server.key"
          http:
            endpoint: 0.0.0.0:4318 # Listen on all interfaces
           # tls:
             # cert_file: "/etc/ssl/certs/collector-server.crt"
             # key_file: "/etc/ssl/private/collector-server.key"
```

The port remains the same.

> If the Collector is publicly available, you can send the telemetry data directly to it. Your Collector should be available under a domain name, for example `collector.example.com:443`. Your web application should send data from its own origin `webapp.example.com` to a different one, and [Cross-Origin Resource Sharing](https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/CORS) (CORS) should be configured so browsers allow sending data to a different origin.

Author: Fixing

        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header Authorization $http_authorization;
        
        proxy_pass https://<motlp-endpoint>:443;
    }
}
```

## Installation

OpenTelemetry packages for web instrumentation are published to npm. You can install them with the package manager of your choice.

The following packages hold the necessary components to set up the base for your instrumentations:

```bash
npm install @opentelemetry/api @opentelemetry/core @opentelemetry/resources @opentelemetry/sdk-trace-base @opentelemetry/sdk-trace-web @opentelemetry/exporter-trace-otlp-http @opentelemetry/instrumentation
```

You can then install the instrumentations that you're interested in. OpenTelemetry offers several instrumentations for browsers. For example:

```bash
npm install @opentelemetry/instrumentation-document-load @opentelemetry/instrumentation-fetch @opentelemetry/instrumentation-xml-http-request @opentelemetry/instrumentation-long-task @opentelemetry/instrumentation-user-interaction
```

## Basic configuration

The minimal configuration you need to instrument your web application with OpenTelemetry includes:

> - **OTEL_EXPORTER_OTLP_ENDPOINT**: The full URL of an OpenTelemetry Collector where data is sent. When using {{product.observability}}, this is the ingest endpoint of an {{serverless-full}} project or the URL of a deployed [EDOT Collector](elastic-agent://reference/edot-collector/index.md). It is likely that the Collector endpoint is of a different origin. If that's the case, you will encounter [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/CORS) issues. Refer to the [OTLP endpoint](#otlp-endpoint) section for more information on how to solve it with different approaches.

Reviewer: Are you referring to ECH with "URL of a deployed EDOT Collector"? Managed OTLP endpoint is available in some regions for ECH. In the [docs](https://www.elastic.co/docs/reference/opentelemetry/motlp#prerequisites) we mention it in the blue "Note" text box.

What I'm thinking about is if we want to mention Managed OLTP endpoint for serverless and ECH plus a deployed collector as a possible OTLP endpoint for the exporter.

- **OTEL_LOG_LEVEL**: Use this configuration to set the log level of the OpenTelemetry components you're going to use.

## Set up OpenTelemetry for the browser

To begin instrumenting your web application with OpenTelemetry in the browser, you need a script. This script configures the essential components, including the context manager, signal providers, processors, and exporters. After setting up the script, you can register the installed instrumentations so they can observe your application and send traces, metrics, and logs to your designated endpoint.

The following start script is in plain JavaScript. If you are using TypeScript, you can adapt this script by changing the file extension to `.ts` and adding the necessary type definitions. OpenTelemetry packages are written in TypeScript, so they include the appropriate type definitions.

:::{note}
Each signal configuration is independent of the others, meaning that you can configure only what you need. The OpenTelemetry API defaults to no-op providers for traces, metrics, and logs.
:::

### Set the configuration

> First, set the configuration options that to be used by all the signals and the instrumentation code. Also initialize the internal logger at the level defined in the configuration.

Reviewer: ```suggestion
First, set the configuration options that are to be used by all the signals and the instrumentation code. Also initialize the internal logger at the level defined in the configuration.
```

    instrumentations: [
      new DocumentLoadInstrumentation(),
      new LongTaskInstrumentation(),
      new FetchInstrumentation(),
      new XMLHttpRequestInstrumentation(),
      new UserInteractionInstrumentation(),
    ],
  });
}
```

## Integrate with your application

With the setup done, it's time to apply it to your web application. You can choose from two main approaches:

1. **Import the code**: Use your build tooling to manage the dependencies and integrate the code into the application bundle. This is the simplest option and is recommended, although it increases the size of your application bundle.

2. **Bundle in a file**: Use a bundler to generate a separate JavaScript file that you include in the `<head>` section of your HTML page. This approach keeps the telemetry code separate from your application bundle.

### Import the code

This approach is recommended. The build tooling manages the dependencies and integrates the code into the application bundle. This might increase the size of your application bundle.

For example, if you're using Webpack, you can import the code like this:

```javascript
// file: app.(js|ts) entry point of your application

> import { initOpenTelemetry } from 'telemetry.js';

Reviewer: Should we mention in the text that we assume the telemetry code is in `telemetry.js`?

Author: Fixing

initOpenTelemetry({
  logLevel: 'info',
  endpoint: 'https://host:port/',
  resourceAttributes: {
    'service.name': 'my-web-app',
    'service.version': '1',
  }
});

// Your app code
```

### Bundle in a file

> You can use a bundler to generate a separate JavaScript file. Place the file within the application's assets folder and include it in the `<head>` section of the HTML page. 

Author: +CC @david-luna

---
navigation_title: OpenTelemetry for Real User Monitoring (RUM)
description: Instrument web applications with OpenTelemetry for Real User Monitoring using Elastic Observability.
applies_to:
  product: preview
  stack:
  serverless:
    observability:
products:
  - id: cloud-serverless
  - id: observability
---

# OpenTelemetry for Real User Monitoring (RUM)

:::{important}
Using OpenTelemetry for Real User Monitoring (RUM) with {{product.observability}} is currently in **Technical Preview** and should not be used in production environments. This feature may be changed or removed in a future release and has [known limitations](#known-limitations). Features in preview are not subject to support SLAs like GA features.
:::

You can instrument your web application with OpenTelemetry browser instrumentation for use with {{product.observability}}. The following sections detail the required components and their proper configuration to acquire traces, logs, and metrics from the application to visualize them within {{kib}}.

## Before you begin [before-you-begin]

> You need a OTLP endpoint to ingest data from the OpenTelemetry RUM instrumentation. If you're setting up a new deployment, [create](/solutions/observability/get-started.md) an {{ecloud}} hosted deployment or {{serverless-short}} project, which includes the [{{motlp}}](opentelemetry://reference/motlp.md). If you own a self-hosted stack or your deployment does not have the {{motlp}}, configure an [EDOT Collector in Gateway mode](https://www.elastic.co/docs/reference/edot-collector/modes#edot-collector-as-gateway).

Author: True! Probably a victim of a hasty rewrite.

        proxy_set_header Authorization 'ApiKey ...your Elastic API key...'; 
        proxy_pass https://collector.example.com:4318;
    }
}
```
:::

:::{warning}
Avoid using the OTel RUM agent alongside any other {{apm-agent}}, including Elastic {{product.apm}} agents. Running multiple agents might lead to conflicting instrumentation, duplicate telemetry, or other unexpected behavior.
:::

## Define the basic settings [otel-rum-basic-settings]

The minimal configuration you need to instrument your web application with OpenTelemetry includes:

- **OTEL_EXPORTER_OTLP_ENDPOINT**: The full URL of the proxy you have configured in the [OTLP endpoint](#before-you-begin) section.
- **OTEL_RESOURCE_ATTRIBUTES**: A JavaScript object that will be used to define the resource. The most important attributes to define are:
  - `service.name` (string): Name of the application you're instrumenting.
  - `service.version` (string, optional): A string representing the version or build of your app.
  - `deployment.environment.name` (string, optional): Name of the environment where the app runs (if applicable); for example, "prod", "dev", or "staging".

> - **OTEL_LOG_LEVEL**: Use this configuration to set the log level of the OpenTelemetry components you're going to use.

Reviewer 3: Do we need available values here (for example `error`, `warn`, `info`, `debug`)?

  exporter: new OTLPMetricExporter({ url: metricsEndpoint }),
});

// Create meter provider to send metrics
const meterProvider = new MeterProvider({
  resource: resource,  // All metrics will be associated with this resource
  readers: [metricReader],
});
metrics.setGlobalMeterProvider(meterProvider);
```
:::
:::::

:::::{step} (Optional) Configure logs

Like traces and metrics, configure a [LoggerProvider](https://opentelemetry.io/docs/specs/otel/logs/sdk/#loggerprovider) if you want to record relevant events that are happening in your application using the API or instrumentations. A LoggerProvider requires several components:

- **Resource**: The resource to be associated with the log records created by the loggers. You defined this in the second step.
- **LogRecord Processor**: A component that manages the log records generated by the loggers and forwards them to a [LogExporter](https://opentelemetry.io/docs/specs/otel/logs/sdk/#logrecordexporter). The exporter should be configured to direct data to an endpoint designated for logs.
- **Log Exporter**: Manages the transmission of log records to the Collector.

For this step, you need the following dependencies:

> - `@opentelemetry/api-logs`: This package contains the logs API. This API is not included yet in the generic API package because logs are still experimental.

Reviewer 3: ```suggestion
- `@opentelemetry/api-logs`: This package contains the logs API. This API is not yet included in the generic API package because logs are still experimental.
```

  
To install the dependencies, run the following command:

```bash
npm install @opentelemetry/instrumentation\
      @opentelemetry/auto-instrumentations-web\
      @opentelemetry/instrumentation-long-task
```

After the dependencies are installed, configure and register instrumentations with the following code:

:::{dropdown} Instrumentations registration
```javascript
import { registerInstrumentations } from '@opentelemetry/instrumentation';
import { getWebAutoInstrumentations } from '@opentelemetry/auto-instrumentations-web';
import { LongTaskInstrumentation } from '@opentelemetry/instrumentation-long-task';

// Register instrumentations
registerInstrumentations({
  instrumentations: [
    getWebAutoInstrumentations(),
    new LongTaskInstrumentation(),
  ],
});
```
:::
:::::

:::::{step} Complete the setup script

> All these pieces together give you a complete setup of all the signals for your web site or application. For convenience, it's better to have it in a separate file which can be named `telemetry.js`. This file should export a function that accepts the configuration allowing you to reuse the setup across different UIs.

Reviewer 3: ```suggestion
All these pieces together give you a complete setup of all the signals for your web site or application. For convenience, it's better to have it in a separate file which can be named, for example, `telemetry.js`. This file should export a function that accepts the configuration allowing you to reuse the setup across different UIs.
```

I presume you can name it something else? If not, then it should be clearly stated.

1. **Import the code**: Use your build tooling to manage the dependencies and integrate the code into the application bundle. This is the simplest option and is recommended, although it increases the size of your application bundle.

2. **Bundle in a file**: Use a bundler to generate a separate JavaScript file that you include in the `<head>` section of your HTML page. This approach keeps the telemetry code separate from your application bundle.

### Import the code 

This approach is recommended and specially if you're using a web framemork. The build tooling manages the dependencies and integrates the code into the application bundle. This approach increases the size of your application bundle.

For example, if you're using Webpack, you can import the code like this:

:::{dropdown} Example: Import telemetry.js in your app

```javascript
import { initOpenTelemetry } from 'telemetry.js';

initOpenTelemetry({
  logLevel: 'info',
  endpoint: 'https://proxy.example.com',
  resourceAttributes: {
    'service.name': 'my-web-app',
    'service.version': '1',
  }
});
```
:::

> If you're using a framework there are some suitable places for it depending on which one you're using:

Reviewer 3: ```suggestion
If you're using a framework, there are some suitable places for it, depending on which one you're using:
```

## Troubleshooting

This section provides solutions to common issues you might encounter when setting up OpenTelemetry for RUM with {{product.observability}}.

:::{dropdown} Module import or bundler errors

If you see errors like "Cannot find module" or bundler-specific issues:

1. Ensure all required packages are installed and listed in `package.json`.

2. Different bundlers (Webpack, Rollup, Vite) may require specific configuration for OpenTelemetry packages.

3. For Webpack, you may need to add polyfills for Node.js modules. Add to your webpack config:

```javascript
resolve: {
  fallback: {
    "process": require.resolve("process/browser"),
    "buffer": require.resolve("buffer/")
  }
}
```

4. For Vite, add to your `vite.config.js`:

```javascript
optimizeDeps: {
  include: ['@opentelemetry/api', '@opentelemetry/sdk-trace-web']
}
```

5. If using TypeScript, ensure `tsconfig.json` has appropriate module resolution:

```json
{
  "compilerOptions": {
    "moduleResolution": "node",
    "esModuleInterop": true,
    "skipLibCheck": true
  }
}
```

:::

:::{dropdown} Reverse proxy configuration issues

If your reverse proxy is not forwarding requests correctly:

> 1. Ensure the reverse proxy (NGINX, Apache, etc.) is running and accessible.

Author: +1

2. Use curl to test the proxy endpoint directly:

```bash
curl -X POST https://your-proxy/v1/traces \
  -H "Content-Type: application/json" \
  -H "Origin: https://your-webapp.example.com" \
  -d '{"test": "data"}' \
  -v
```

3. Review proxy logs for errors or blocked requests.

4. Ensure the proxy can reach the backend EDOT Collector or mOTLP endpoint.

:::

:::{dropdown} Configuration issues

If your OpenTelemetry setup isn't initializing correctly:

1. Ensure `OTEL_EXPORTER_OTLP_ENDPOINT` doesn't include the signal path (like `/v1/traces`). The script provided adds this automatically:

```javascript
// Correct
const OTEL_EXPORTER_OTLP_ENDPOINT = 'https://collector.example.com:4318';

// Incorrect
const OTEL_EXPORTER_OTLP_ENDPOINT = 'https://collector.example.com:4318/v1/traces';
```

2. Verify `service.name` is set and doesn't contain special characters:

```javascript
const OTEL_RESOURCE_ATTRIBUTES = {
  'service.name': 'my-web-app', // Required
  'service.version': '1.0.0',
};
```

3. Ensure providers are registered before instrumentations:

```javascript
// Correct order:
// 1. Configure and register tracer provider

> trace.setGlobalTracerProvider(tracerProvider);

Reviewer 2: ```suggestion
tracerProvider.register({ contextManager: new ZoneContextManager() });
```

:::{warning}
Avoid using OpenTelemetry alongside any other {{apm-agent}}, including Elastic {{product.apm}} agents. Running multiple agents in the same application process might lead to conflicting instrumentation, duplicate telemetry, or other unexpected behavior.
:::

### OTLP endpoint

You have two main options for setting up an OTLP endpoint:

1. Use an existing OpenTelemetry Collector: If you already have a Collector deployed in your infrastructure, you can configure it to accept telemetry from browsers. This requires configuring CORS headers either directly on the Collector (if it's publicly accessible) or through a reverse proxy.

2. **Start from scratch with {{ecloud}}**: If you're setting up a new deployment, you can create an {{ecloud}} hosted deployment or {{serverless-short}} project, which includes the [{{motlp}}](opentelemetry://reference/motlp.md). This approach requires a reverse proxy to handle CORS configuration.

#### Use an existing OpenTelemetry Collector [use-an-existing-open-telemetry-collector]

> If you already have an OpenTelemetry Collector in your infrastructure, you can reuse it to ingest the traces, metrics, and logs generated by your instrumented web applications.

Reviewer 4: It's a bit unclear from this description (apart from the CORS config) where the "existing Collector" would send the data? 

- (a) i.e. would it at the same time serve as a gateway collector (i.e. aggregate trace metrics and send data directly to Elasticsearch using the ES exporter)? 
- (b) OR do we recommend that to be separate from a Gateway collector and it would send data through OTLP to another "Gateway collector" / mOTLP?

If it's (a) we should clearly state that it needs to be an EDOT-like collector (with all the required components included for metrics aggregation) with a corresponding config for metrics aggregation. If it's (b) we should state it explicitly as well.

IMHO, in the case of ECH and Serverless, we should rather recommend (b) as it simplifies things.

Author: +CC @hegerchr @david-luna

```nginx
server {
    listen 80 default_server;
    server_name _;
    location / {
        # Take care of preflight requests
        if ($request_method = 'OPTIONS') {
            add_header 'Access-Control-Max-Age' 1728000;
            add_header 'Access-Control-Allow-Origin' 'webapp.example.com' always;
            add_header 'Access-Control-Allow-Headers' 'Accept,Accept-Language,Content-Language,Content-Type,Authorization' always;
            add_header 'Access-Control-Allow-Credentials' 'true' always;
            add_header 'Content-Type' 'text/plain charset=UTF-8';
            add_header 'Content-Length' 0;
            return 204;
        }

        add_header 'Access-Control-Allow-Origin' 'webapp.example.com' always;
        add_header 'Access-Control-Allow-Credentials' 'true' always;
        add_header 'Access-Control-Allow-Headers' 'Accept,Accept-Language,Content-Language,Content-Type,Authorization' always;
        proxy_http_version 1.1;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;

>         proxy_set_header Authorization $http_authorization;

Reviewer 2: This config is proxying the `Authorization` header from the browser most likely in the form `ApiKey ...`. I have my doubts if we want the key being configured in the web app where is public.

        
        proxy_pass https://<motlp-endpoint>:443;
    }
}
```

:::

## Installation

OpenTelemetry packages for web instrumentation are published to npm. You can install them with the package manager of your choice.

The following packages hold the necessary components to set up the base for your instrumentations:

```bash
npm install @opentelemetry/api @opentelemetry/core @opentelemetry/resources @opentelemetry/sdk-trace-base @opentelemetry/sdk-trace-web @opentelemetry/exporter-trace-otlp-http @opentelemetry/instrumentation
```

You can then install the instrumentations that you're interested in. OpenTelemetry offers several instrumentations for browsers. For example:

> - **OTEL_EXPORTER_OTLP_ENDPOINT**: The full URL of an OpenTelemetry Collector where data is sent. When using {{product.observability}}, this is the {{motlp}} of an {{serverless-full}} project or the URL of a deployed [EDOT Collector](elastic-agent://reference/edot-collector/index.md). Refer to the [OTLP endpoint](#otlp-endpoint) section for more information on how to solve [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/CORS) issues.

Reviewer 2: I think this leaked from the config section. It should be removed

To define the resource, install the following dependencies:

- `@opentelemetry/resources`: This package helps you to define and work with resources because a Resource is not a plain object and has some properties (like immutability) and constraints.
- `@opentelemetry/browser-detector`: Detectors help you to define a resource by querying the runtime and environment and resolving some attributes. In this case, the browser detector resolves the language, brands, and mobile attributes of the browser namespace.

To install the dependencies, run the following command:

```bash
npm install @opentelemetry/resources @opentelemetry/browser-detector
```

After the dependencies are installed, define the resource for your instrumentation with the following code:

```javascript
import { resourceFromAttributes, detectResources } from '@opentelemetry/resources';
import { browserDetector } from '@opentelemetry/opentelemetry-browser-detector';

> // Append this in the code section

Reviewer 2: ```suggestion
```

:::

This documentation outlines the process for instrumenting your web application with OpenTelemetry browser instrumentation, using {{product.observability}} as the backend. Unlike the [EDOT SDKs](opentelemetry://reference/edot-sdks/index.md), this approach uses upstream OpenTelemetry JavaScript packages directly. The following sections detail the required components and their proper configuration to acquire traces, logs, and metrics from the application to visualize them within {{kib}}.

While this guide uses upstream OpenTelemetry instrumentation, you can also use the [EDOT Collector](elastic-agent://reference/edot-collector/index.md) components as part of your data ingestion pipeline.

## Prerequisites

This guide assumes you're using an {{product.observability}} deployment. You can use an existing one or set up a new one. If you're new to {{product.observability}}, follow the guidelines in [Get started with {{product.observability}}](/solutions/observability/get-started.md).

:::{warning}

> Avoid using OpenTelemetry alongside any other {{apm-agent}}, including Elastic {{product.apm}} agents. Running multiple agents in the same application process might lead to conflicting instrumentation, duplicate telemetry, or other unexpected behavior.

Reviewer 5: ```suggestion
Avoid using OTel RUM agent alongside any other {{apm-agent}}, including Elastic {{product.apm}} agents. Running multiple agents in the same application process might lead to conflicting instrumentation, duplicate telemetry, or other unexpected behavior.
```

:::

### OTLP endpoint

> You need an OTLP Collector to ingest data from the OpenTelemetry RUM instrumentation. If you're setting up a new deployment, you can create an {{ecloud}} hosted deployment or {{serverless-short}} project, which includes the [{{motlp}}](opentelemetry://reference/motlp.md).

Reviewer 6: ```suggestion
You need an OpenTelemetry Collector to receive data from the OpenTelemetry RUM instrumentation. If you're setting up a new deployment, you can create an {{ecloud}} hosted deployment or {{serverless-short}} project, which includes the [{{motlp}}](opentelemetry://reference/motlp.md).
```

Depeding on where the collector is placed in your infrastructure you might have two setup options:

1. Reverse proxy (recommended): Take this approach if you're using {{motlp}} or if your collector is not publicly available. Refer to the next section for further information.

2. Setup Cross-Origin Resource Sharing (CORS): You can opt in for this setup if your collector is public and you can modify its configuration.

::::{tab-set}

:::{tab-item} Reverse proxy (recommended)

Use a reverse proxy to redirect the requests from your web app to the collector for these reasons:

> - There is no option in {{motlp}} to configure [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/CORS) and without the proper configuration browsers can't send data to it.

Reviewer 5: Yes its enabled by default in mOTLP HTTP endpoint.

This is a basic EDOT Collector configuration file that activates CORS:

```yaml
receivers:
  # Receives data from other Collectors in Agent mode or OTEL SDKs
  otlp:
    protocols:
      grpc:
        endpoint: 0.0.0.0:4317 # Listen on all interfaces
        auth:
          authenticator: apikeyauth
      http:
        endpoint: 0.0.0.0:4318 # Listen on all interfaces
        auth:
          authenticator: apikeyauth
        cors: # Configure CORS for RUM. ref: https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/confighttp/README.md#server-configuration
          allowed_origins:
            - http://*.example.com
            - https://*.example.com
connectors:
  elasticapm: {} # Elastic APM Connector

processors:
  batch:
    send_batch_size: 1000
    timeout: 1s
    send_batch_max_size: 1500
  batch/metrics:
    send_batch_max_size: 0 # Explicitly set to 0 to avoid splitting metrics requests
    timeout: 1s
  elastictrace: {} # Elastic Trace Processor

exporters:
  debug: {}

>   elasticsearch/otel:

Reviewer 5: Shouldnt we be recommending our mOTLP endpoint via OTLP exporter?

Reviewer 4: That's related to my original comment / question: what role do we want this collector to play and what is our opinionated recommendation? 

To make it concrete, would there be any downsides / limitations if we would ONLY document the reverse proxy setup here and omit describing the collector setup (but rather refer to existing docs that do that already for mOTLP / Gateway in case of self-managed):

- The architectural differences for Serverless, ECH and self-managed are already covered [here (and corresponding sub-pages)](https://www.elastic.co/docs/reference/opentelemetry/architecture). 
- The only RUM-specific part is how to receive data from publicly running RUM SDKs --> that what the reverse proxy would solve
- So, we only describe how to setup the proxy and forward that OTLp data to either mOTLP (for ECH and Serverless) or the Gateway collector (in the case of self-managed). But we don't need to cover the collector config for the gateway collector here, just refer to the self-managed setup docs. 

Than we could get rid of this entire collector config here and also of additional options. I think that would simplify these docs.

(Unless there are downsides with the proxy-approach that are addressed by the public collector setup isntead).

To install the dependencies, run the following command:

```bash
npm install @opentelemetry/sdk-metrics @opentelemetry/exporter-metrics-otlp-http
```

After the dependencies are installed, configure and register a meter provider with the following code:

:::{dropdown} Meter provider configuration

```javascript
import { metrics } from '@opentelemetry/api';
import { MeterProvider, PeriodicExportingMetricReader } from '@opentelemetry/sdk-metrics';
import { OTLPMetricExporter } from '@opentelemetry/exporter-metrics-otlp-http';

// Set the metrics endpoint based on the config provided
const metricsEndpoint = `${OTEL_EXPORTER_OTLP_ENDPOINT}/v1/metrics`;

// Create metric reader to process metrics and export using OTLP
const metricReader = new PeriodicExportingMetricReader({
  exporter: new OTLPMetricExporter({ url: metricsEndpoint }),
});

// Create meter provider to send metrics
const meterProvider = new MeterProvider({
  resource: resource,  // All metrics will be associated with this resource
  readers: [metricReader],
});
metrics.setGlobalMeterProvider(meterProvider);
```

:::

:::::

:::::{step} Configure logs

> For RUM log management with OpenTelemetry JavaScript, configure the **Provider** for generation (instantiation, resource, logger creation) and the **Exporter** for transmission (endpoint, headers, interval/batching, registration).

Reviewer 5: This section can be improved as well.

  const tracerProvider = new WebTracerProvider({
    resource,
    spanProcessors: [
      new BatchSpanProcessor(new OTLPTraceExporter({
        url: tracesEndpoint,
      })),
    ],
  });
  trace.setGlobalTracerProvider(tracerProvider);

  // Metrics signal setup
  const metricsEndpoint = `${config.endpoint}/v1/metrics`;
  const metricReader = new PeriodicExportingMetricReader({
    exporter: new OTLPMetricExporter({ url: metricsEndpoint }),
  });
  const meterProvider = new MeterProvider({
    resource: resource,
    readers: [metricReader],
  });
  metrics.setGlobalMeterProvider(meterProvider);

  // Logs signal setup
  const logsEndpoint = `${config.endpoint}/v1/logs`;
  const logExporter = new OTLPLogExporter({ url: logsEndpoint });

  const loggerProvider = new LoggerProvider({
    resource: resource,
    processors: [new BatchLogRecordProcessor(logExporter)]
  });
  logs.setGlobalLoggerProvider(loggerProvider);

  // Register instrumentations
  registerInstrumentations({
    instrumentations: [
      new DocumentLoadInstrumentation(),
      new LongTaskInstrumentation(),
      new FetchInstrumentation(),
      new XMLHttpRequestInstrumentation(),

>       new UserInteractionInstrumentation(),

Reviewer 2: https://github.com/open-telemetry/opentelemetry-js-contrib/pull/702 was closed. Will eventually appear in https://github.com/open-telemetry/opentelemetry-browser but it is not released yet
